File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head==NULL) return NULL;
        int len=0;
        ListNode* ptr=head;
        while(ptr){
            ++len;
            ptr=ptr->next;
        }
        if(k>len) return head;
        int t=k;
        ListNode *prev=NULL;
        ptr=head;
        while(ptr && --t>=0){
            ListNode *next=ptr->next;
            ptr->next=prev;
            prev=ptr;
            ptr=next;
        }
        head->next=reverseKGroup(ptr,k);
        return prev;
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        int size=0;
        ListNode ptr=head;
        while(ptr!=null){
            ptr=ptr.next;
            ++size;
        }
        if(k>size) return head;
        int temp=k;
        ListNode prev=null;
        ptr=head;
        // ListNode next;
        while(temp>0 && ptr!=null){
            ListNode next=ptr.next;
            ptr.next=prev;
            prev=ptr;
            ptr=next;
            --temp;
        }
        head.next=reverseKGroup(ptr,k);
        return prev;
    }
}